core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1749,6 +1752,9 @@ class Log(@volatile private var _dir: File,
           checkIfMemoryMappedBufferClosed()
           // remove the segments for lookups
           removeAndDeleteSegments(deletable, asyncDelete = true, reason)
+          if (reason == LogCompaction) {
+            maybeIncrementLogStartOffset(segments.firstEntry.getValue.baseOffset, SegmentCompaction)
+          }
           maybeIncrementLogStartOffset(segments.firstEntry.getValue.baseOffset, SegmentDeletion)
```

Comment:
Shouldn't we be passing `reason` instead of what we're doing here?

Reply:
Looks like the types are not the same, but this approach seems error prone.

Reply:
Sorry if this is a little confusing. `reason` is for deletion and the other parameter `SegmentCompaction` is the reason for changing the offset. Currently that is used to handle the issue for updating past the high watermark. I agree that it is not the cleanest way and we should figure out what we want to do with compacting past the high watermark.

Reply:
I've removed this now that I've changed to not clean past the high watermark.